services/config.go

```diff
@@ -0,0 +1,264 @@
+package services
+
+import (
+	"bytes"
+	"compress/gzip"
+	"errors"
+	"fmt"
+	"io"
+	"net/http"
+	"net/url"
+	"strconv"
+	"sync"
+	"time"
+
+	"github.com/getlantern/detour"
+	"github.com/getlantern/golog"
+	"google.golang.org/protobuf/proto"
+	"google.golang.org/protobuf/types/known/timestamppb"
+
+	"github.com/getlantern/flashlight/v7/apipb"
+	"github.com/getlantern/flashlight/v7/common"
+	"github.com/getlantern/flashlight/v7/ops"
+)
+
+const (
+	defaultConfigSaveDir  = ""
+	defaultConfigFilename = "proxies.conf"
+
+	defaultConfigPollInterval = 3 * time.Minute
+	defaultConfigPollJitter   = 2 * time.Minute
+)
+
+// ConfigOptions specifies the options to use for ConfigService.
+type ConfigOptions struct {
+	// URL to use for fetching this config.
+	OriginURL string
+
+	// UserConfig contains data for communicating the user details to upstream
+	// servers in HTTP headers, such as the pro token and other options.
+	UserConfig common.UserConfig
+
+	// Sticky specifies whether or not to only use the local config and not
+	// update it with remote data.
```

Comment:
> Not sure if this comment was ported from somewhere else, but I find the near double negative confusing.
Yeah, that is the original comment. 

> Actually, this guarantee is not possible with your current API. You don't have any control in this package over the first result of ConfigHandler.GetConfig (since it's an interface with no strict requirements on the source of the first call).

If `Sticky` is true, `configService` returns [here](https://github.com/getlantern/flashlight/pull/1395/files/f7de44c45aecd7155b6b96f0ae9c1fbdacfd22af#diff-6edc66a61226196eb1a3035bd1dad77b4c22d2cab11dc33ae62c9cfa4c71a8e9R137-R139) and doesn't start the loop of fetching the remote config. `configService` only fetches remote configs and be an interface between `ConfigHandler` and other packages. Since `ConfigHandler` is responsible for the actual config, `configService` assumes that the initial value has already be set. `proxyconfig.Init` reads in the local config file and sets it as the initial value for `proxyconfig.config` [here](https://github.com/getlantern/flashlight/pull/1395/files/f7de44c45aecd7155b6b96f0ae9c1fbdacfd22af#diff-be33908ff3493e3cab2744eb648322f15e8af73624cac4cbac593caa1473f24bR59-R64).

Maybe renaming `Sticky` to something like `FetchRemote` would make it much more clear.

Also, initially, I considered implementing it such that `configService`'s only purpose was to fetch the remote config and update it as needed and have other packages call `proxyconfig` directly to get the config (since this is the only interaction between other packages and `configService`). This way, if `Sticky` is true then you just don't start `configService`. I don't remember why I decided not to go that route, but maybe that is the solution. What do you think?

Reply:
@hwh33 and I discussed this on a call and decided that it would be better to move the responsibility of returning the config to other packages to `proxyconfig`, resulting in `configService`'s only responsibility is to fetch new configs form the server and relaying that to `proxyconfig` accordingly.